Adds a default color

kjbenk opened the pull request:
Allows developers to supply a new option `default_color` to specify whether or not the color picker will use the clear or default functionality.

From https://github.com/alleyinteractive/wordpress-fieldmanager/issues/490

kjbenk commented:
@mboynes Please have a quick look at the regex I am using to make sure the default color is set.

srtfisher commented:
Curious about this: what is your rational for not using the `default_value` property?

kjbenk commented:
@srtfisher I am not sure I understand your question.  This PR is meant to allow developers to add a default color for the color picker itself.  This will change the `Clear` button to a `Default` button.

bcampeau commented:
I think this should use `default_value`

kjbenk commented:
@srtfisher I understand what you are referring to now 😄 

https://github.com/alleyinteractive/wordpress-fieldmanager/blob/master/php/class-fieldmanager-field.php#L233-L237

mboynes commented on tests/php/test-fieldmanager-colorpicker-field.php:
```diff
@@ -27,6 +27,15 @@ public function test_basic_render() {
+		ob_start();
+		$fm->add_meta_box( 'Test Colorpicker', 'post' )->render_meta_box( $this->post, array() );
+		$html = ob_get_clean();
+		$this->assertEquals( 1, preg_match( '[data-default-color="#ff0000"]', $html ) );
```
You can use `$this->assertRegExp()` (see the test just above this one)

mboynes replied:
while technically valid, square brackets are pretty uncommon for regex delimiters, especially since they're special characters in regex. The typically delimiter is `/`, except when that character is used in the regex, then you'll commonly find `#` used.

All that said, make this `'/data-default-color="#ff0000"/'`

mboynes commented:
The `$default_value` question is a good one and I had to think a lot about it. The question is: would you ever want the default in the color picker to be different from the initial value? There is one circumstance I can think of, which is where you want the value to be blank initially, but still be able to provide a default. This would especially be valid in repeating fields/groups, where setting a default value on a field will cause an otherwise empty field/group to save (e.g. when `extra_elements = 1`, which would cause a new field/group to be added with every save of the screen).

Here's my suggestion: Add a new variable `$default_color` as you did before. If `$default_value` is set and `$default_color` is not, set `$default_color` to `$default_value`. Be sure to _not_ set `$default_color` to anything, and use `isset()` to check it, that way someone could deliberately set the default color to an empty string if they wanted.

kjbenk commented:
@mboynes I implemented the method you have described above and added unit tests for each scenario.  Please let me know if that is what you envisioned.

mboynes commented on php/class-fieldmanager-colorpicker.php:
```diff
@@ -43,6 +50,12 @@ public function __construct( $label = '', $options = array() ) {
 
 		$this->sanitize = array( $this, 'sanitize_hex_color' );
 
+		if ( ! empty( $options['default_color'] ) ) {
```
make this `isset()`

kjbenk replied:
My thinking was that since there is no real reason for a `''` default_color value, we should instead check for `! empty()` and not `isset()` to account for that.  In other words, when the user sets a default_color of `''` then we should change it to the default_value.  I hope that makes sense 😄

mboynes commented on tests/php/test-fieldmanager-colorpicker-field.php:
```diff
@@ -27,6 +27,32 @@ public function test_basic_render() {
+		ob_start();
+		$fm_3->add_meta_box( 'Test Colorpicker', 'post' )->render_meta_box( $this->post, array() );
+		$html = ob_get_clean();
+		$this->assertRegExp( '/data-default-color/', $html );
```
Should be:
```
$this->assertRegExp( '/data-default-color=""/', $html );
```

mboynes commented on tests/php/test-fieldmanager-colorpicker-field.php:
```diff
@@ -27,6 +27,32 @@ public function test_basic_render() {
+		ob_start();
+		$fm_4->add_meta_box( 'Test Colorpicker', 'post' )->render_meta_box( $this->post, array() );
+		$html = ob_get_clean();
+		$this->assertRegExp( '/data-default-color/', $html );
```
Same as above

mboynes commented on tests/php/test-fieldmanager-colorpicker-field.php:
```diff
@@ -27,6 +27,32 @@ public function test_basic_render() {
+		ob_start();
+		$fm_2->add_meta_box( 'Test Colorpicker', 'post' )->render_meta_box( $this->post, array() );
+		$html = ob_get_clean();
+		$this->assertRegExp( '/data-default-color="#ff0000"/', $html );
```
May as well also check value while you're at it...

```php
$this->assertRegExp( '/data-default-color="#ff0000" value="#ff0000"/', $html );
```

I guess there's no reason not to add this to the others, too

mboynes commented:
:+1:
